### PR TITLE
image_repository: size the colour-matrix copy from the blob, not the field

### PR DESCRIPTION
**Pull request by mmoqui**
During the investigation about the crash of Ansel at its starting (see PR #1374), Claude IA found a potential bug when the size of the color matrix isn't at the expected size.

image_repository: size the colour-matrix copy from the blob, not the field

dt_image_from_stmt() read main.images.color_matrix with

    if(color_matrix) memcpy(img->d65_color_matrix, color_matrix,
                            sizeof(img->d65_color_matrix));

which trusts the row to hold nine floats because the destination does. dt_image_repository_store() does bind exactly that many, so every row this schema wrote is fine -- but a library inherited from darktable need not be, and a shorter blob walks up to 36 bytes off sqlite's own buffer.

Copy only when the blob is the full 3x3, and restore the documented "no matrix" sentinel otherwise. The second half matters as much as the bounds check: every consumer tests isnan(d65_color_matrix[0]) (imageio_profile.c twice, imageio_rgbe.c), and dt_image_init() only memsets the struct, so leaving the field untouched on a short blob would hand them an all-zero matrix that reads as perfectly valid and gets fed to dt_colorspaces_create_xyzimatrix_profile().

math.h is included explicitly for NAN rather than left to arrive through one of the other headers, per the include rule in CLAUDE.md.

Found while chasing an unrelated heap corruption; no reproducer, since producing one needs a pre-Ansel database. It is a latent out-of-bounds read, not the cause of that crash.

### CODE REVIEW

File: src/database/image_repository.c [hunk 1]
```diff
@@ -136,8 +137,19 @@ static void dt_image_from_stmt(dt_image_t *img, sqlite3_stmt *stmt)
     uint32_t tmp = sqlite3_column_int(stmt, 34);
     memcpy(&img->legacy_flip, &tmp, sizeof(dt_image_raw_parameters_t));
   }
+  /* Size the copy from the blob, not from the destination. The writer binds exactly
+   * sizeof(d65_color_matrix) (see dt_image_repository_store()), so a short blob means a row this
+   * schema did not write -- a library inherited from darktable -- and reading nine floats out of
+   * it walks off sqlite's buffer. A partial matrix is not a usable one either, so anything that
+   * is not the full 3x3 restores the documented "no matrix" sentinel instead: every consumer
```
Comment (Jiyone) [inline src/database/image_repository.c:144]: **Line 144-146**:

The comment is wrong on both of its claims, and the same two sentences are in the commit message.

**"leaving the field at the calloc'd zeroes"** — `dt_image_init()` does more than `memset`: `common/image.c:1972` explicitly sets `img->d65_color_matrix[0] = NAN;`. On the allocation path (`caches/image_cache.c:237-238`: `dt_image_init()`, then `_image_cache_reload_from_db()`) the field was therefore never all-zero, and this `else` branch changes nothing there. What it actually protects is the **reload** path: `dt_image_cache_get_reload()` (`caches/image_cache.c:381-390`) reloads into an already populated struct **without** going through `dt_image_init()`, so without the branch a matrix from a previous load would survive a row that no longer carries one. Good reasoning — just not the one written down.

**"every consumer (imageio_profile.c, imageio_rgbe.c) tests isnan"** — `imageio_rgbe.c:655` is a *producer* (`mat3inv()` writes *into* the field), not a reader. The readers left out are `iop/colorin.c:1017` and `pixel/illuminants.h:417`; and `common/image.c:278` goes through `_image_matrix_has_data()` (isfinite + non-zero sum), which already rejects both NaN and all-zero correctly.

Suggested rewrite, replacing lines 140-146:

```suggestion
  /* Size the copy from the blob, not from the destination: the writer binds exactly
   * sizeof(d65_color_matrix) (see dt_image_repository_store()), so anything shorter is a row
   * this schema did not write -- a library inherited from darktable -- and reading nine floats
   * out of it is undefined. The else branch is the one that runs in practice, on the NULL
   * column of a row not yet stored: dt_image_cache_get_reload() reloads into an already
   * populated struct without going through dt_image_init(), so without it a matrix from a
   * previous load would survive a row that no longer carries one. NAN is the documented
   * "no matrix" sentinel every reader tests (imageio_profile.c, iop/colorin.c,
   * pixel/illuminants.h); zeroes there would read as a perfectly valid matrix. */
```

File: src/database/image_repository.c [hunk 2]
```diff
@@ -136,8 +137,19 @@ static void dt_image_from_stmt(dt_image_t *img, sqlite3_stmt *stmt)
...
+   * (imageio_profile.c, imageio_rgbe.c) tests isnan(d65_color_matrix[0]), and leaving the field
+   * at the calloc'd zeroes would hand them an all-zero matrix that reads as valid. */
   const void *color_matrix = sqlite3_column_blob(stmt, 35);
-  if(color_matrix) memcpy(img->d65_color_matrix, color_matrix, sizeof(img->d65_color_matrix));
+  const int color_matrix_size = sqlite3_column_bytes(stmt, 35);
+  if(color_matrix && color_matrix_size == (int)sizeof(img->d65_color_matrix))
```
Comment (Jiyone) [inline src/database/image_repository.c:149]: `IS_NULL_PTR()`, and the null test is redundant anyway.

AGENTS.md requires `IS_NULL_PTR()` / `!IS_NULL_PTR()` for every null-pointer test. More to the point, the test buys nothing here: sqlite returns `NULL` only for a NULL or zero-length value, so `color_matrix_size == 36` already implies a non-null pointer. One condition is enough, and the intermediate variable goes away with it.

Replacing lines 147-150:

```suggestion
  const int color_matrix_size = sqlite3_column_bytes(stmt, 35);
  if(color_matrix_size == (int)sizeof(img->d65_color_matrix))
    memcpy(img->d65_color_matrix, sqlite3_column_blob(stmt, 35), sizeof(img->d65_color_matrix));
```

The strict `==` rather than `>=` is the right call, though: a longer blob is not a 3x3 this code knows how to read.

File: src/database/image_repository.c [hunk 3]
```diff
@@ -136,8 +137,19 @@ static void dt_image_from_stmt(dt_image_t *img, sqlite3_stmt *stmt)
     uint32_t tmp = sqlite3_column_int(stmt, 34);
     memcpy(&img->legacy_flip, &tmp, sizeof(dt_image_raw_parameters_t));
   }
+  /* Size the copy from the blob, not from the destination. The writer binds exactly
+   * sizeof(d65_color_matrix) (see dt_image_repository_store()), so a short blob means a row this
+   * schema did not write -- a library inherited from darktable -- and reading nine floats out of
+   * it walks off sqlite's buffer. A partial matrix is not a usable one either, so anything that
```
Comment (Jiyone) [inline src/database/image_repository.c:143]: "walks off sqlite's buffer" overstates it — measured, that is not what happens.

A minimal sqlite program under ASAN (16-byte blob, read with the pre-fix 36-byte `memcpy`), tried in both regimes — inline record, then a row forced onto overflow pages — reports **no** error: the value lives inside sqlite's page buffer, or inside an over-allocated `Mem`. The concrete effect is a garbage matrix, not an out-of-allocation read.

It is still UB by API contract — nothing in sqlite promises that headroom — and the fix stands as written; the commit message just sells a buffer overrun that is not observable. `reading nine floats out of it is undefined` is enough, and is already folded into the rewrite above.

**Review by Jiyone — CHANGES_REQUESTED** [review]

File: src/database/image_repository.c (hunk 1 above)
Comment (mmoqui) [reply src/database/image_repository.c]: Both corrections confirmed, and both applied in 171fef1b01 (amended, not stacked — the commit message carried the same two wrong claims).

`dt_image_init()` does set the sentinel: `common/image.c:1972`. My earlier grep stopped short of it in a function longer than I assumed, and I built the "calloc'd zeroes" argument on that miss. The field is never all-zero on the allocation path, so the `else` branch changes nothing there.

Your replacement reasoning checks out too: `dt_image_cache_get_reload()` (`caches/image_cache.c:392`) calls `_image_cache_reload_from_db()` on an already populated struct with no `dt_image_init()` in between, so a matrix from a previous load would survive a row that no longer carries one. I also checked the branch cannot eat a live matrix: `dt_image_cache_write_release()` persists it through `dt_image_repository_store()` (which binds the full 36 bytes) before any reload, so it only fires when the row genuinely has none.

Consumer list corrected to `imageio_profile.c`, `iop/colorin.c:1017` and `pixel/illuminants.h:417`. `imageio_rgbe.c:655` is indeed a producer. `dt_image_is_matrix_correction_supported()` goes through `_image_matrix_has_data()` and is mentioned in the commit message as already handling both cases.

I took your rewrite with one change: I dropped the closing "zeroes there would read as a perfectly valid matrix". Once `dt_image_init()` is in the picture, zeroes never occur on either path — what the reload leaves behind is a real matrix from the previous image, which is the actual hazard. Keeping the zeroes clause would have carried a smaller version of the same false claim.

File: src/database/image_repository.c (hunk 2 above)
Comment (mmoqui) [reply src/database/image_repository.c]: Agreed on the null test, and applied — with one deliberate difference from the suggestion.

You are right that it buys nothing: sqlite returns NULL only for a NULL or zero-length value, so a size of exactly 36 already says the pointer is good. Dropping it removes the `IS_NULL_PTR()` question entirely rather than answering it, and the intermediate variable goes with it.

What I did not take is the call order. The suggestion puts `sqlite3_column_bytes()` before `sqlite3_column_blob()`, and `sqlite3.h:5236` documents the opposite as the safe one:

> The safest policy is to invoke these routines in one of the following ways: [...] sqlite3_column_blob() followed by sqlite3_column_bytes() [...] you should call sqlite3_column_text(), sqlite3_column_blob(), or sqlite3_column_text16() first to force the result into the desired format, then invoke sqlite3_column_bytes()

A size asked for first is allowed to describe a value the later call then converts. For a column this schema wrote it makes no difference — it is already a blob — but the whole premise of this patch is a row written by something else, so I would rather not lean on the value's type being what we expect. So:

```c
  const void *color_matrix = sqlite3_column_blob(stmt, 35);
  if(sqlite3_column_bytes(stmt, 35) == (int)sizeof(img->d65_color_matrix))
    memcpy(img->d65_color_matrix, color_matrix, sizeof(img->d65_color_matrix));
  else
    img->d65_color_matrix[0] = NAN;
```

The comment now states that the order is deliberate, so nobody tidies it back the other way. Happy to switch if you read the sqlite contract differently — it is the one point here I am not certain enough about to call settled.

And yes, `==` rather than `>=` was the intent: a longer blob is not a 3x3 this code knows how to read.

File: src/database/image_repository.c (hunk 3 above)
Comment (mmoqui) [reply src/database/image_repository.c]: Accepted — the overstatement was mine, and your measurement is the kind of thing I should have run before writing "walks off sqlite's buffer" rather than after being told.

Both the comment and the commit message now claim only what the API contract gives: reading nine floats out of a shorter blob is undefined, and the observable effect is a garbage matrix. The commit message records your ASAN result explicitly — 36-byte read of a 16-byte blob, inline record and overflow pages both, no error — so the next reader does not have to re-derive that the buffer overrun is not the point. The fix stands on the contract alone.
